cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandlerInternal.java

```
@@ -200,4 +219,72 @@ public void getAppDetailForVersion(HttpRequest request, HttpResponder responder,
         : applicationLifecycleService.getAppDetail(appId);
     responder.sendJson(HttpResponseStatus.OK, GSON.toJson(appDetail));
   }
+
+  /**
+   * Deploy an application.
+   *
+   * @param request {@link HttpRequest}
+   * @param responder {@link HttpResponder}
+   * @param namespaceId of the namespace where the app is to be deployed
+   * @param appId of the app
+   * @param skipMakingLatest if true, the app will be deployed but not marked latest
+   * @return {@link BodyConsumer}
+   * @throws BadRequestException when the request params or body are not valid
+   * @throws NamespaceNotFoundException when the namespace is not found
+   * @throws AccessException in case of any security issues
+   */
+  @PUT
+  @Path("/apps/{app-id}")
+  @AuditPolicy(AuditDetail.REQUEST_BODY)
+  public BodyConsumer create(HttpRequest request, HttpResponder responder,
+      @PathParam("namespace-id") final String namespaceId,
+      @PathParam("app-id") final String appId,
+      @QueryParam("skipMakingLatest") final boolean skipMakingLatest)
+      throws BadRequestException, NamespaceNotFoundException, AccessException {
+    String versionId = ApplicationId.DEFAULT_VERSION;
+    if (Feature.LIFECYCLE_MANAGEMENT_EDIT.isEnabled(featureFlagsProvider)) {
+      versionId = RunIds.generate().getId();
+    }
+    ApplicationId applicationId = validateApplicationVersionId(validateNamespace(namespaceId), appId, versionId);
+
+    try {
+      return deployAppFromArtifact(applicationId, tmpDir, "internalAppRequest-",skipMakingLatest);
+    } catch (Exception ex) {
```

don't need to catch this, let it propagate up.

---

done